src/test/java/com/squareup/javapoet/TypeNameTest.java

```diff
@@ -99,4 +100,14 @@ private void assertEqualsHashCodeAndToString(TypeName a, TypeName b) {
     assertThat(a.hashCode()).isEqualTo(b.hashCode());
   }
 
+  @Test public void binary() {
+    assertEquals("void", TypeName.VOID.binary());
+    assertEquals("int", TypeName.INT.binary());
+    assertEquals("java.lang.Object", TypeName.OBJECT.binary());
```

Comment:
What specification are you trying to implement here? This looks a little like the JVM’s field descriptor, but that spec uses a `/` instead of a `.`.

http://docs.oracle.com/javase/specs/jvms/se7/html/jvms-4.html#jvms-4.3